### ISSUE TITLE
A Question about YOLOv4Head output shape

### ISSUE THREAD

**Issue by Junhojuno** [argusswift/YOLOv4-pytorch#195]
Hello! 
I have a Question about the codes in yolo_head.
```python
# in /model/head/yolo_head.py /
def forward(self, p):
        bs, nG = p.shape[0], p.shape[-1]
        p = p.view(bs, self.__nA, 5 + self.__nC, nG, nG).permute(0, 3, 4, 1, 2)
        p_de = self.__decode(p.clone())
        return (p, p_de)
...

def __decode(self, p):
        ...
        pred_xy = (torch.sigmoid(conv_raw_dxdy) + grid_xy) * stride        
        pred_wh = (torch.exp(conv_raw_dwdh) * anchors) * stride        
        pred_xywh = torch.cat([pred_xy, pred_wh], dim=-1)       
        pred_conf = torch.sigmoid(conv_raw_conf)        
        pred_prob = torch.sigmoid(conv_raw_prob)        
        pred_bbox = torch.cat([pred_xywh, pred_conf, pred_prob], dim=-1)

```
The outputs of yolo_head are p, p_de.
p is used to compute confidence & class loss, the others using p_de.
I was wondering if I could use the p_de for calculating the confidence loss & class prob loss.
Is there any purpose to use both of those, not just p_de.

**Comment by argusswift** [comment]
> Hello!
> I have a Question about the codes in yolo_head.
> 
> ```python
> # in /model/head/yolo_head.py /
> def forward(self, p):
>       bs, nG = p.shape[0], p.shape[-1]
>       p = p.view(bs, self.__nA, 5 + self.__nC, nG, nG).permute(0, 3, 4, 1, 2)
>       p_de = self.__decode(p.clone())
>       return (p, p_de)
> ...
> 
> def __decode(self, p):
>       ...
>       pred_xy = (torch.sigmoid(conv_raw_dxdy) + grid_xy) * stride        
>       pred_wh = (torch.exp(conv_raw_dwdh) * anchors) * stride        
>       pred_xywh = torch.cat([pred_xy, pred_wh], dim=-1)       
>       pred_conf = torch.sigmoid(conv_raw_conf)        
>       pred_prob = torch.sigmoid(conv_raw_prob)        
>       pred_bbox = torch.cat([pred_xywh, pred_conf, pred_prob], dim=-1)
> ```
> 
> The outputs of yolo_head are `p`, `p_de`.
> `p` is used to compute confidence & class loss, the others using `p_de`.
> I was wondering if I could use the `p_de` for calculating the confidence loss & class prob loss.
> Is there any reason to use both of those(`p`, `p_de`), not just `p_de`.

p is used for calculating model loss and p_de is ready for geting test results(box 、confidence and class)
